### PR TITLE
feat: Refactor LLM model zoo and add KV cache support

### PR DESCRIPTION
**Pull request by peri044**
# Description

This PR redesigns our LLM model compilation, unifies it, fixes output mismatch and performance issues. This PR also implements KV caching using native TensorRT. 

Fixes # (issue)

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: tools/llm/cache_utils.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+from torch._export.utils import _detect_fake_mode_from_gm
+from torch._ops import OpOverloadPacket
+from torch._subclasses.fake_tensor import FakeTensor, FakeTensorMode
+from torch.fx import Graph, GraphModule, Node
+from torch.fx.node import Target
+from torch.fx.passes.shape_prop import _extract_tensor_metadata
+from torch.utils._pytree import _LEAF_SPEC
+
+
+@torch_tensorrt.dynamo.conversion.dynamo_tensorrt_converter(
```
Comment (narendasan) [inline tools/llm/cache_utils.py]: Do we want to upstream this eventually?

File: tools/llm/sdpa_converter.py [hunk 2]
```diff
@@ -62,25 +62,15 @@ def scaled_dot_product_attention(
 ) -> TRTTensor:
```
Comment (narendasan) [inline tools/llm/sdpa_converter.py]: I would put the torch_trt extensions in a subdirectory all together. like //tools/llm/torch_trt_ext/

File: tools/llm/test_qwen2.5_components.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
+import torch
```
Comment (narendasan) [inline tools/llm/test_qwen2.5_components.py:1]: Should we create a `tests/py/tools/llm`?

File: tools/llm/cache_utils.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+    return random_tensors
+
+
+def add_graph_input(
+    gm: GraphModule, name: str, val: Optional[torch.Tensor] = None, dynamic_shape=None
+) -> Node:
+    """Add a graph input to the given GraphModule and return the newly created node.
+
+    NOTE: function does NOT do any graph canonicalization. This is left to the user!
```
Comment (narendasan) [inline tools/llm/cache_utils.py:110]: If you dont want users to hit this directly consider prefixing the method name with `_` like `_add_graph_input`

File: tools/llm/cache_utils.py (hunk 1 above)
Comment (peri044) [reply tools/llm/cache_utils.py]: eventually yes. But maybe not for this release.

File: tools/llm/test_qwen2.5_components.py (hunk 3 above)
Comment (peri044) [reply tools/llm/test_qwen2.5_components.py:1]: We can. These tests helped me debug accuracy/functionality issues during the development. It would take some time to get them working in CI. I can add them as a part of cherry-pick PR after this. What do you think ?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 5]
```diff
@@ -837,6 +859,7 @@ def contains_metadata(gm: torch.fx.GraphModule) -> bool:
             continue
         submodule_node_dict[node.name] = node
 
+    preserve_module_specs(original_in_spec, original_out_spec, partitioned_module)
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/_compiler.py:869]: Should set spec for `partitioned_module`'s children?

File: examples/dynamo/weight_streaming_example.py [hunk 6]
```diff
@@ -32,7 +32,43 @@
 import torch
 import torch_tensorrt
 from transformers import AutoModelForCausalLM
-from utils import export_llm
+
+
+def export_llm(model, inputs, min_seq_len=1, max_seq_len=16):
```
Comment (zewenli98) [inline examples/dynamo/weight_streaming_example.py:37]: why not directly call tools/llm/utils.py?

**Review by zewenli98 — COMMENTED** [review]
Mostly LGTM. just some questions

File: examples/dynamo/weight_streaming_example.py (hunk 6 above)
Comment (peri044) [reply examples/dynamo/weight_streaming_example.py:37]: We have to use sys.append to include the path to tools/llm/utils.py to do that. I wanted this example to be standalone so I just copied this function.

File: tools/llm/cache_utils.py (hunk 1 above)
Comment (peri044) [reply tools/llm/cache_utils.py]: removed it now

File: tools/llm/sdpa_converter.py (hunk 2 above)
Comment (peri044) [reply tools/llm/sdpa_converter.py]: done

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 5 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/_compiler.py:869]: The spec for children is actually set by the partitioning. I'm only trying to preserve the output spec for the entire partitioned_module (It should be the same as original module unless we change it explicitly which is the case in KV caching)

File: tools/llm/utils.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        start_time = timeit.default_timer()
+        _ = generate_fn(model, inputs, output_seq_length, eos_token_id)
+        torch.cuda.synchronize()
+        end_time = timeit.default_timer()
+        timings.append(end_time - start_time)
+
+    return timings
+
+
+def recordStats(backend, timings, precision, batch_size=1, compile_time_s=None):
```
Comment (narendasan) [inline tools/llm/utils.py]: Super minor but everything else is snake case

**Review by narendasan — APPROVED** [review]
LGTM only suggestion is moving tests into a tests subdirectory or into the tests directory. Is optional.
